tests/subsys/bluetooth/services/ble_bas/src/unity_test.c

```diff
@@ -0,0 +1,450 @@
+
+	bas_cfg.can_notify = false;
+	bas_init(&bas_cfg);
+	__cmock_sd_ble_gatts_value_set_ExpectAnyArgsAndReturn(NRF_ERROR_INVALID_PARAM);
+	ret = ble_bas_battery_level_update(&ble_bas, conn_handle, battery_level);
+	TEST_ASSERT_EQUAL(-EINVAL, ret);
+
+	bas_cfg.can_notify = true;
+	bas_init(&bas_cfg);
+	__cmock_sd_ble_gatts_value_set_ExpectAnyArgsAndReturn(NRF_SUCCESS);
+	__cmock_sd_ble_gatts_hvx_ExpectAnyArgsAndReturn(NRF_ERROR_TIMEOUT);
+	ret = ble_bas_battery_level_update(&ble_bas, conn_handle, battery_level);
+	TEST_ASSERT_EQUAL(-EINVAL, ret);
+}
+
+void test_ble_bas_battery_level_update_enotconn(void)
+{
+	int ret;
+	uint16_t conn_handle = 0x0001;
+	uint32_t battery_level = 42;
+	struct ble_bas_config bas_cfg = bas_cfg_template;
+
+	bas_cfg.can_notify = true;
+	bas_init(&bas_cfg);
+
+	__cmock_sd_ble_gatts_value_set_ExpectAnyArgsAndReturn(NRF_SUCCESS);
+	__cmock_sd_ble_gatts_hvx_ExpectAnyArgsAndReturn(BLE_ERROR_INVALID_CONN_HANDLE);
+	ret = ble_bas_battery_level_update(&ble_bas, conn_handle, battery_level);
+	TEST_ASSERT_EQUAL(-ENOTCONN, ret);
+}
+
+void test_ble_bas_battery_level_update_epipe(void)
+{
+	int ret;
+	uint16_t conn_handle = BLE_CONN_HANDLE_INVALID;
+	uint32_t battery_level = 42;
+	struct ble_bas_config bas_cfg = bas_cfg_template;
+
+	bas_cfg.evt_handler = ble_bas_evt_handler;
+	bas_init(&bas_cfg);
+	__cmock_sd_ble_gatts_value_set_ExpectAnyArgsAndReturn(NRF_SUCCESS);
+	__cmock_sd_ble_gatts_hvx_ExpectAnyArgsAndReturn(NRF_ERROR_INVALID_STATE);
+	ret = ble_bas_battery_level_update(&ble_bas, conn_handle, battery_level);
+	TEST_ASSERT_EQUAL(-EPIPE, ret);
+	TEST_ASSERT_EQUAL(battery_level, ble_bas.battery_level);
+}
+
+static uint32_t stub_sd_ble_gatts_hvx_param_check(uint16_t conn_handle,
+						  ble_gatts_hvx_params_t const *p_hvx_params,
+						  int num_calls)
+{
+	TEST_ASSERT_NOT_NULL(p_hvx_params);
+	TEST_ASSERT_EQUAL(VALUE_HANDLE, p_hvx_params->handle);
+	TEST_ASSERT_EQUAL(BLE_GATT_HVX_NOTIFICATION, p_hvx_params->type);
+	TEST_ASSERT_EQUAL(0, p_hvx_params->offset);
+	TEST_ASSERT_EQUAL(sizeof(uint8_t), *(p_hvx_params->p_len));
+	TEST_ASSERT_EQUAL(ble_bas.battery_level, *(uint8_t *)(p_hvx_params->p_data));
+	return NRF_SUCCESS;
+}
+
+void test_ble_bas_battery_level_update_success(void)
+{
+	int ret;
+	uint16_t conn_handle = 0x0001;
+	uint32_t battery_level = 55;
+	struct ble_bas_config bas_cfg = bas_cfg_template;
+
+	bas_cfg.can_notify = false;
+	bas_init(&bas_cfg);
+
+	/* Battery level hasn't change 'Nothing to do' */
+	ret = ble_bas_battery_level_update(&ble_bas, conn_handle, battery_level);
+	TEST_ASSERT_EQUAL(0, ret);
+	TEST_ASSERT_EQUAL(battery_level, ble_bas.battery_level);
+
+	/* Change battery level, and ble_bas should update but not notify */
+	battery_level = 42;
+	__cmock_sd_ble_gatts_value_set_ExpectAnyArgsAndReturn(NRF_SUCCESS);
+	ret = ble_bas_battery_level_update(&ble_bas, conn_handle, battery_level);
+	TEST_ASSERT_EQUAL(0, ret);
+	TEST_ASSERT_EQUAL(battery_level, ble_bas.battery_level);
```

Comment:
Instead of accepting any arguments to `sd_ble_gatts_value_set()` you should check that the implementation is passing the value that the user provided when calling `ble_bas_battery_level_updated()`. The implementation may pass one  value to `sd_ble_gatts_value_set()` and a set different value in `ble_bas.battery_level` (due to a bug).